### ISSUE TITLE
investigate TLS13 on MacOS

### ISSUE THREAD

**Issue by wfurt** [dotnet/runtime#1979]
It seems like on Catalina TLS1.3 works out of the box - at least with Safari. 
However, when I try to use in .NET I get the following error:
```
System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception.
 ---> Interop+AppleCrypto+SslException: bad protocol version
   --- End of inner exception stack trace ---
```

This may be because API we currently use is marked as obsolete or there may be something missing since TLS1.3 support got out before Catalina. And all TLS1.3 tests are disabled on all versions of MacOS. 

related to https://github.com/dotnet/runtime/issues/1720

**Comment by wfurt** [comment]
It seems like we will need to rewrite PAL to use new Network framework API.
That should fix iOS as well.

https://mailarchive.ietf.org/arch/msg/tls/5QjzTilqjomSyzENtgfaAqQOhbA/

**Comment by karelz** [comment]
Triage: Will require PAL rewrite on Mac -- out of scope of 5.0.
The new Catalina API will help also iOS and light up TLS 1.3 there.

**Comment by chucker** [comment]
Is this still current? In a .NET 6 console app with macOS 12.3, I still get the same stacktrace as above. .NET info:

<details>
.NET SDK (reflecting any global.json):
 Version:   6.0.201
 Commit:    ef40e6aa06

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  12.3
 OS Platform: Darwin
 RID:         osx.12-x64
 Base Path:   /usr/local/share/dotnet/sdk/6.0.201/

Host (useful for support):
  Version: 6.0.3
  Commit:  c24d9a9c91

.NET SDKs installed:
  6.0.201 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.3 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
</details>

**Comment by wfurt** [comment]
yes, it is @chucker. #27727 has more detail but it is essentially same work. It is not clear at this moment if it is even feasible.

**Comment by bhoffman-splunk** [comment]
Not to add pressure, but I am more and more encountering TLS 1.3-only systems. This is especially true for internal, greenfield APIs that companies build, since they can "start secure" from the beginning. Any work around (such leveraging OpenSSL as is done in Linux) would be tremendously helpful.

**Comment by wfurt** [comment]
Is there need for client, server or both? I was thinking about some fallback because to allow HTTP2 for macOS developers (https://github.com/dotnet/runtime/issues/27727) We generally do that for HTTP3/Quic (preview in 6.0) so we have some parts already in a place. _if_ we go down this path, it would probably be opt-in feature with dependency on OpenSSL from Brew/MacPorts. Would that be acceptable?

**Comment by chucker** [comment]
@wfurt client in my case. macOS as a server is uncommon. But server might still be useful (if less important) for developers who run Kestrel on a Mac?

**Comment by wfurt** [comment]
And yes, the server part would be for Mac developers (like me). Production deployment of Kestrel on macOS is unlikely IMHO. 

Just to elaborate more since the use case here is different from  https://github.com/dotnet/runtime/issues/27727.

Aside testing and support, the biggest challenge with using OpenSSL on macOS is integration with keychain and certificate management. Using Apple's AP allows to use private keys that are non-exportable and private keys are not in core dumps. And Apple packages LibreSSL instead of OpenSSL and that is not binary compatible. So dependency management is going to be hard. 

The new Apple's api makes it very difficult (and maybe impossible) for `SslStream` to consume. Based on testing @filipnavara did, client may be easier and there may some possibilities. The challenge would be moving forward with client if there is not solution for the server. That would effectively mean we would need to maintain both PAL layers.

**Comment by bhoffman-splunk** [comment]
@wfurt sorry for not being more clear. This is for clients.

As a client, not being able to make encrypted connections to TLS 1.3-only services/servers is by far the bigger issue. I'm finding companies start a new green-field API or service, and configure it to only use TLS 1.3. Writing .NET code that can talks to services configured like this will only grow with time.

At this point I'm literally shelling out from .NET to run "ncat + openssl" via `Process` class to make requests for me and get the response, which is super clumsy and slow.

**Comment by wfurt** [comment]
ok. thanks for the feedback @bhoffman-splunk and @chucker. I'm not sure if this fits to 7.0 but I'll try to take another look and I'll sync up with with @filipnavara.

**Comment by filipnavara** [comment]
I am gonna be away for the next week but happy to discuss it when I get back.

**Comment by davidkaya** [comment]
Hi, are there any plans on supporting TLS 1.3 on macOS in future .NET releases? 

Our usecase is that we have customers with 10s of thousands of users using our software (.NET 6) on their macOS. The server side is deployed on a Windows Server which has only TLS 1.3 enabled. The software deployed on workstations currently cannot communicate with the server side due to the TLS 1.3 requirement.

**Comment by wfurt** [comment]
triage: we should investigate for 8.0

**Comment by filipnavara** [comment]
Here's a link to the example on how to "misuse" the framer API in Network.framework to perform client-side TLS: https://gist.github.com/filipnavara/d5fb55bdb5edcceb1981f73078b855c4

(I'll be happy to answer any questions about it once I get back from holiday at the beginning of December.)

**Comment by lini** [comment]
Hi there, 
It seems that TLS1.3 support for macOS clients is not coming in the .NET 8.0 release this year. Currently we have to use a third party library (Bouncy Castle) to workaround this on our mac clients, but it is not a perfect solution - e.g. the performance penalty for high network throughput threads is high. 
What is the current status of this feature - is the implementation possible at all using the Apple native APIs?

**Comment by wfurt** [comment]
It is on 9.0 wish list for me. We simply run out of time in 8.0 and it is not trivial effort. Based on prototypes from @filipnavara it should be doable at least for the client side. While not ideal I would be OK for me doing only half for now - just to make some progress. ALPN was like that for a while until https://github.com/dotnet/runtime/issues/27727 plugged server side in 8.0. But in the mean time HttpClient (and others) could connect to HTTP/2 servers.

**Comment by rzikm** [comment]
Triage: big effort, unclear implementation direction (the new API does not make it easy to get TLS messages in a buffer so that we can push them to the provided inner stream ourselves). But would be great for platform compatibility.

**Comment by elias-ik** [comment]
Its 2024, and we still can't call TSL1.3 APIs?
Any plans?

**Comment by jozefizso** [comment]
We are writing troubleshooting utility in .NET to verify our app can correctly connected using IPv4 and IPv6 and using different TLS protocols.

On macOS the app reports only TLS v1.2 connection.

Will TLS 1.3 really be ready in .NET 9?

**Comment by wfurt** [comment]
It is still high on the priority list. However, it is not clear if it is even doable because of  Apple's new AI shape.

**Comment by wfurt** [comment]
you would need TLS terminating proxy @pmauchle. It is probably doable but tricky. I really wish we can make progress but realistically we don't have resources to get it done in 9.

**Comment by stvndall** [comment]
Azure services are now highlighting that TLS 1.2 needs to be upgraded to TLS 1.3. After a day of not understanding why local development calling to hosted services was suddenly not working, we found the security team upgraded based on Azure's recommendation. Half the team ended up with issues. 

Worth noting our Win 11 team members had the same issues. It might be a similar issue. 
This really needs a fix, considering Azure is recommending to upgrade, and our security team says they can select either 1.2 or 1.3, not both. Considering both are Microsoft (all 3 if win 11 is broken for the same reason), this needs a patch before dotnet can not run on security recommendations.

**Comment by liveans** [comment]
Hello @stvndall,
This is something we're actively working on it right now, and we're targeting to finish this in January.

**Comment by karelz** [comment]
@stvndall can you tell us more about your Mac usage and why are your devs blocked? What are the apps you are using that require only TLS 1.3?
Note that we plan to fix this in upcoming release - .NET 10. The change required to support TLS 1.3 on Mac is a non-trivial feature and we won't be able to backport it into .NET 9.
In the meantime, we would recommend to keep your Azure services on TLS 1.2, which will enable your Mac workloads. The Azure warning is IMHO a bit aggressive on security recommendations and there is not immediate danger with TLS 1.2 being deprecated or compromised security-wise.

Just FYI: The key reason why TLS 1.3 on Mac is not supported yet by .NET, is because Apple decided to not implement it in their low-level APIs (where they support all previous TLS versions). They exposed only new high-level APIs which encapsulate also TCP connection. That model is not working for .NET, because we expose the TCP connection under TLS. Therefore we have to do a lot of hackary around the new Apple APIs, which is complicated.

**Comment by wfurt** [comment]
I'm wondering where the recommendation is coming from @stvndall? AFAIK the recommendation for Azure is 1.2+ -> and that is what all our services use. 

Also from the description it seems like you really need just the client side, right? e.g. functionality from #104835 would be sufficient? It is tricky business and we will be looking for beta testers at some point. Hopefully it will land in 10.0.

**Comment by johnnypea** [comment]
This is ridiculous 🙄

**Comment by tskardal** [comment]
I might be missing some pieces here, so sorry in advance if that is. I too am unable to develop against an API that requires TLS 1.3.

> I'm wondering where the recommendation is coming from [@stvndall](https://github.com/stvndall)? AFAIK the recommendation for Azure is 1.2+ -> and that is what all our services use.

Requiring TLS 1.3 for Azure App Services is at least recommended by Microsoft Defender for Cloud as a _Medium risk level_.

> Also from the description it seems like you really need just the client side, right? e.g. functionality from [#104835](https://github.com/dotnet/runtime/pull/104835) would be sufficient? It is tricky business and we will be looking for beta testers at some point. Hopefully it will land in 10.0.

When writing clients for HTTP+JSON APIs you'll end up with this kind of error:

```csharp
using System.Security.Authentication;

var handler = new HttpClientHandler
{
    // Results in the following exception:
    // Unhandled exception. System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
    // ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception.
    // ---> System.PlatformNotSupportedException: The requested security protocol is not supported.
    SslProtocols = SslProtocols.Tls13
    
    // This will work (if the server accepts it):
    // SslProtocols = SslProtocols.Tls12
};

using var client = new HttpClient(handler);
var response = await client.GetAsync("https://github.com");
response.EnsureSuccessStatusCode();
var content = await response.Content.ReadAsStringAsync();
Console.WriteLine(content);
```

**Comment by jogibear9988** [comment]
Isn't it possible that netcore uses openssl on other os ?
in https://github.com/dotnetprojects/S7CommPlusDriver/tree/master/src/S7CommPlusDriver/runtimes we use it for ssl, but we have to deliver openssl for every OS in our package. This could be done my the MS netcore installer for each system

**Comment by rzikm** [comment]
@jogibear9988 we want to specifically avoid shipping OpenSSL binaries. Shipping crypto code is always problematic due to various certifications and restrictions. Also delivery of critical security fixes is more difficult for crypto code which is not part of the given platform (e.g. Schannel fixes may be delivered as part of Windows Update, but your deployed application which shipped with openssl library has to be patched manually).

**Comment by wfurt** [comment]
One there are challenges for integration with KeyChains. The scenario @tskardal described would be covered by https://github.com/dotnet/runtime/pull/104835 or some of its reincarnations. The problematic part is server side e.g. https://github.com/dotnet/aspnetcore/issues/60200

Whoever is anxious can either contribute or at least try to convince Apple to provide useable API.

**Comment by dwarfland** [comment]
Are there any plans after 5 years to have this addressed and support TLS 1.3 on macOS? More and more servers are changing to require 1.3 (and the error reporting also doesn't make ti very clear what the issue is, when it fails).

**Comment by tverboon** [comment]
I created an Apple Feedback issue on 21 March: `FB16957804` referencing this issue and provided comments from this issue. Status is `Open`.

**Comment by alkhimey** [comment]
Has anyone been able to find a workaround for a client situation (we have a client connecting to a server that allows only TLS 1.3). Would be happy to hear details on how you managed to make it work.

**Comment by MattParkerDev** [comment]
@alkhimey 

You can proxy via ghostunnel locally, and let it handle TLS 1.3
https://github.com/ghostunnel/ghostunnel

`ghostunnel client --listen localhost:1337 --target remoteurl.com:443  --disable-authentication`

Note that you may get errors back regarding an incorrect host header, you'll need to override the default Host header like this:

```csharp
var client = new HttpClient();
client.DefaultRequestHeaders.Host = "remoteurl.com";
```

**Comment by chucker** [comment]
> we want to specifically avoid shipping OpenSSL binaries.

Is the code that ultimately links the TLS library tightly coupled, or would it be feasible for third parties to swap in a different SSL library while still using `HttpClient`, etc.?

**Comment by rzikm** [comment]
Experimental TLS 1.3 Client support has been merged in https://github.com/dotnet/runtime/pull/117428. It will ship in .NET 10 preview 7

**Comment by rzikm** [comment]
Moving to 11.0.0 milestone for server implementation and further evolution.

**Comment by wfurt** [comment]
We probably should share instructions and limitations are for 10. As this is opt-in preview-ish feature .... but it should get most of the use cases unblocked.

**Comment by rzikm** [comment]
We will prepare some release notes, cc @liveans

**Comment by liveans** [comment]
In .NET 10 Preview 7, we've added TLS 1.3 Client-side only support for OSX. You can find details about how to enable it.

**Quick Start:**

Enable via AppContext switch:
```csharp
AppContext.SetSwitch("System.Net.Security.UseNetworkFramework", true);
```

Or via environment variable:
```bash
DOTNET_SYSTEM_NET_SECURITY_USENETWORKFRAMEWORK=1
```

**Key Details:**
- **Opt-in only** – existing apps continue using the current stack
- macOS client-side only in this release
- Older TLS versions (1.0/1.1) may not be available when enabled
- Applies to `SslStream` and dependent APIs like `HttpClient`

**Important Differences:**
- Stream behavior may differ (buffering, read/write, cancellation)
- Zero-byte read semantics may change
- Some IDN hostnames may be rejected – use ASCII/Punycode when possible
- Cipher suites are controlled by macOS

We appreciate your feedbacks, we'll keep our eye on this and keep improving it in further releases.

Release notes: https://github.com/dotnet/core/blob/main/release-notes/10.0/preview/preview7/libraries.md#tls-13-for-macos-client

**Comment by PedroCavaleiro** [comment]
This partially worked I now ran into other issue when using the `Microsoft.EntityFrameworkCore.SqlServer` if this switch is enabled either by AppContext or Environment Variable the HttpClient works just fine but I run into issues with the SQL Connection

```
Microsoft.Data.SqlClient.SqlException (0x80131904): A connection was successfully established with the server, but then an error occurred during the login process. (provider: TCP Provider, error: 35 - An internal exception was caught)
 ---> System.IO.IOException: The read operation failed, see inner exception.
 ---> Interop+NetworkFramework+NetworkFrameworkException: Network Framework error -9836 in domain TLS
   at System.Net.Security.SafeDeleteNwContext.<FillAppDataBufferAsync>g__CompletionCallback|42_0(IntPtr context, NetworkFrameworkError* error, Byte* data, Int32 length)
--- End of stack trace from previous location ---
   at System.Net.Quic.ResettableValueTaskSource.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Threading.Tasks.ValueTask.ValueTaskSourceAsTask.<>c.<.cctor>b__4_0(Object state)
--- End of stack trace from previous location ---
   at System.Net.Security.SafeDeleteNwContext.<ReadAsync>g__ReadAsyncInternal|41_0(Memory`1 buffer, CancellationToken cancellationToken)
   at System.Net.SyncReadWriteAdapter.WaitAsync[T](ValueTask`1 task)
   at System.Net.Security.SslStream.ReadAsyncInternal[TIOAdapter](Memory`1 buffer, CancellationToken cancellationToken), ErrorCode: -9836, ErrorDomain: TLS
   --- End of inner exception stack trace ---
```

Trying to find a fix but so far no luck, removing the switch makes the DB accessible again but the HttpClient that I rely on stops working

* macOS 26.1 (25B78)
* .NET SDK 10.0.100
* Microsoft.EntityFrameworkCore.SqlServer 10.0.0

**Comment by wfurt** [comment]
Can you do packet capture @PedroCavaleiro ? The error is `errSSLPeerProtocolVersion` and I'm wondering if this is because the SQL server supports only some very old TLS version. We should be able to see this from the handshake exchange.

**Comment by PedroCavaleiro** [comment]
Will do, but for context I use Azure SQL database

**Comment by wfurt** [comment]
Do you know if there is any TLS configuration for the DB access? This is are I'm not very familiar with. But with using Azure it should be easier for us to reproduce and investigate.

**Comment by PedroCavaleiro** [comment]
My sys admin told me that this specific database it's not yet migrated to Azure SQL it's running in a VM 

The screenshot is the packet capture and I can spot the usage of TLS 1.2

<img width="1078" height="451" alt="Image" src="https://github.com/user-attachments/assets/06f4e37f-b1a7-42a5-8863-6b2584f53126" />

<img width="685" height="674" alt="Image" src="https://github.com/user-attachments/assets/be10b1d7-d53b-4f21-a846-62a6308a7342" />

If you need a export of a specific packet let me know

-- UPDATE --

I uploaded the packet dissections and obfuscated the IPs but 10.0.0.1 = SQL Server and 10.0.0.2 = macOS 

[packet-dissections.json](https://github.com/user-attachments/files/23547042/packet-dissections.json)

**Comment by wfurt** [comment]
I don't see anything obvious. Client offered several versions and server picked 1.2 with some EC suite. Then it has some exchange and it dies. We will probably need to set up some repro. I don't think we have done any testing with SQL. There is another layer of complexity where the TLS frames are chipped and embedded inside TDS.

**Comment by fhenriquezn** [comment]
Hey guys, I am experiencing the same TLS protocol error on .NET 10 with macOS Tahoe (26.1) on Apple M4 Max. When running `dotnet restore --verbosity detailed`, it fails with: 

The SSL connection could not be established...
Authentication failed...
bad protocol version

I've already tried `dotnet dev-certs https --trust` and clearing local certificates, but the issue persists. Is there a specific workaround or fix for this environment?

**Comment by wfurt** [comment]
Are you restoring from some private server or is it reachable publicly @fhenriquezn? @liveans has repro for the SQL case above and he is trying to figure it out.

**Comment by fhenriquezn** [comment]
@wfurt  Yes, it's a private server. It's behind Cloudflare. I had the "Minimum TLS Version" setting configured to 1.3 in the Cloudflare dashboard (SSL/TLS → Edge Certificates).

Changing it back to 1.2 allowed dotnet restore to succeed immediately.

To clarify, my goal is to use TLS 1.3. I was under the impression that native support (which wasn't slated for .NET 9) would arrive with .NET 10, but that doesn't seem to be fully the case yet.

**Comment by wfurt** [comment]
It is not enabled by default @fhenriquezn. See https://github.com/dotnet/runtime/issues/1979#issuecomment-3417929998
Current goal is to flip the switch in 11.0 if we can resolve reported issues.

**Comment by fhenriquezn** [comment]
Yes, but that will only partially solve it. That configuration may work for the running application, but not for CLI commands like `dotnet restore`—am I correct?

**Comment by wfurt** [comment]
Using `DOTNET_SYSTEM_NET_SECURITY_USENETWORKFRAMEWORK=1` environment should work of any tool.

**Comment by fhenriquezn** [comment]
That worked, thank you very much :)

**Comment by wfurt** [comment]
Just to be sure: you set the MinTLS back to 1.3 and everything is good so far, right?

**Comment by fhenriquezn** [comment]
Yes sir 

<img width="2560" height="1440" alt="Image" src="https://github.com/user-attachments/assets/675594f6-70b5-4192-b39f-5b7073ea7477" />

1. I am using zsh, so `nano ~/.zshrc` 
   1.2. add `export DOTNET_SYSTEM_NET_SECURITY_USENETWORKFRAMEWORK=1` then save
3. I also added it into my Program.cs as commented before:
```
if (OperatingSystem.IsMacOS())
{
    AppContext.SetSwitch("System.Net.Security.UseNetworkFramework", true);
}
